### ISSUE TITLE
[Android] Slow screen opening after react-native update

### ISSUE THREAD

**Issue by OrlovMaks** [th3rdwave/react-native-safe-area-context#448]
Hey, everybody!
After updating react-native (and all libraries) from version 0.70.11 to 0.72.6, I got a problem that screens with SafeAreaView open slowly and without animation.
I couldn't find absolutely no information on how to fix it.

In the video the first screen "Security" is in SafeAreaView, the second screen "Your rank" is just View.
https://github.com/th3rdwave/react-native-safe-area-context/assets/93343100/45623a6f-a038-4372-837a-abfa7466b8e8

My App component is wrapped with <SafeAreaProvider initialMetrics={initialWindowMetrics}> (I tried with and without the initialMetrics parameter).

My package.json
{ "react": "18.2.0",
"react-native": "0.72.6",
"react-native-safe-area-context": "4.7.4",
"react-native-screens": "3.27.0",
"react-native-reanimated": "3.5.4",
"@react-navigation/bottom-tabs": "6.5.11",
"@react-navigation/native": "6.1.9",
"@react-navigation/stack": "6.3.20",

**Comment by efstathiosntonas** [comment]
Maybe a long shot but, do you by any chance have reduce motion enabled on the device/simulator?

**Comment by 17Amir17** [comment]
I see this as-well something like a 500ms delay on rendering whatever is inside safe view

**Comment by jacobp100** [comment]
There's not really anything to go off here. If you manage to make a minimal reproduction - or better yet, find the actual issue - let us know.

**Comment by OrlovMaks** [comment]
@jacobp100 @17Amir17 
I prepared a repository with the project from which I removed everything except the structure of my navigation:
https://github.com/OrlovMaks/sample_safeAreaView_issue

While I was preparing this project I found a problem (check the screenshot):
I have detachInactiveScreens={false} everywhere and I make navigation from WalletStackNavigator to general StackNavigator if I remove detachInactiveScreens from this general stack everything works.
https://github.com/th3rdwave/react-native-safe-area-context/assets/93343100/4225fb1f-275d-4252-aefc-ac4d0ad50316

**Comment by jacobp100** [comment]
Right so is it a problem with this library or another?

**Comment by 17Amir17** [comment]
I have this issue with [BottomSheetModal](https://gorhom.github.io/react-native-bottom-sheet/modal). I created https://github.com/17Amir17/SafeAreaContext to reproduce

https://github.com/th3rdwave/react-native-safe-area-context/assets/36531255/56a39589-92a9-4971-9a2b-d9967909dc22

I'm not really sure how to debug this further but would love to help @jacobp100 if you have any directions

**Comment by mgerdes** [comment]
@jacobp100 I'm seeing this issue too. It seems like it's because is we are hitting this timeout in SafeAreaView and that freezes the UI for 500ms

https://github.com/th3rdwave/react-native-safe-area-context/blob/b086b0ef5043836b72c6430cca1641641694fbce/android/src/main/java/com/th3rdwave/safeareacontext/SafeAreaView.kt#L97-L99

I debugged a little bit and it seems like the code in `runOnNativeModulesQueueThread` below is never able to acquire the lock, because the lock is being held the entire time we are waiting in the while loop right below, so it doesn't complete until the timeout ends.

https://github.com/th3rdwave/react-native-safe-area-context/blob/b086b0ef5043836b72c6430cca1641641694fbce/android/src/main/java/com/th3rdwave/safeareacontext/SafeAreaView.kt#L77-L95

I'm not sure what the best fix here is though.

**Comment by jacobp100** [comment]
Good catch on that. If you're able to get some flame graph and figure out what exactly is blocking the main thread, that would be helpful. Is this something that shows up only in development - or production too?

**Comment by mgerdes** [comment]
> Good catch on that. If you're able to get some flame graph and figure out what exactly is blocking the main thread, that would be helpful. Is this something that shows up only in development - or production too?

I think the `waitForReactLayout` in `SafeAreaView.kt` is what's blocking the main thread. I added these log statements to try to see what's happening 

<img width="606" alt="Screenshot 2023-11-01 at 1 18 43 PM" src="https://github.com/th3rdwave/react-native-safe-area-context/assets/7636272/e61b2e4c-fba7-4fe9-ad4b-6df596eb4081">

and this was the result:

```
2023-11-01 13:03:54.865 runOnNativeModulesQueueThread 0
2023-11-01 13:03:54.865 main thread 0
2023-11-01 13:03:54.866 awaitNanos 0
2023-11-01 13:03:55.367 awaitNanos 1
2023-11-01 13:03:55.367 main thread 1
2023-11-01 13:03:55.367 Timed out waiting for layout.
2023-11-01 13:03:55.377 native thread 0
2023-11-01 13:03:55.377 native thread 1
```

So if I'm looking at this right it seems like the code in `runOnNativeModulesQueueThread` that's supposed to run on the native thread isn't run while the main thread is blocked and that makes `waitForReactLayout` block the main thread the entire time.

**Comment by jacobp100** [comment]
If you increase the max time nanos to something much longer (say 10s), does it finish before? Or do we have a deadlock?

**Comment by mgerdes** [comment]
> If you increase the max time nanos to something much longer (say 10s), does it finish before? Or do we have a deadlock?

Yea, I just upped it to 10s and the whole app freezes for those 10 seconds it's waiting. So yea, something isn't letting that code run on the native thread.

**Comment by jacobp100** [comment]
What’s your setup? Version of react native? Are you using fabric?

**Comment by mgerdes** [comment]
We are on react-native 0.72.5. No we are not using fabric. 

I printed the value of `getReactContext(this).isOnNativeModulesQueueThread` at the start of the function and outside of  `runOnNativeModulesQueueThread`, and it returns false both times.

**Comment by jacobp100** [comment]
Try adding a `printStackTrace` in the `runOnNativeModulesQueueThread` function in the react native library code and see if there’s something else that runs around that time. It sounds like this and another task are in deadlock

**Comment by mgerdes** [comment]
Here's what the stacktrace looks like in `runOnNativeModulesQueueThread`

```
java.lang.Exception: Stack trace
	at java.lang.Thread.dumpStack(Thread.java:1615)
	at com.th3rdwave.safeareacontext.SafeAreaView.waitForReactLayout(SafeAreaView.kt:71)
	at com.th3rdwave.safeareacontext.SafeAreaView.updateInsets(SafeAreaView.kt:52)
	at com.th3rdwave.safeareacontext.SafeAreaView.maybeUpdateInsets(SafeAreaView.kt:113)
	at com.th3rdwave.safeareacontext.SafeAreaView.onAttachedToWindow(SafeAreaView.kt:134)
	at android.view.View.dispatchAttachedToWindow(View.java:21357)
	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3491)
	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3498)
	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3498)
	at android.view.ViewGroup.addViewInner(ViewGroup.java:5291)
	at android.view.ViewGroup.addView(ViewGroup.java:5077)
	at com.facebook.react.views.view.ReactViewGroup.addView(ReactViewGroup.java:516)
	at android.view.ViewGroup.addView(ViewGroup.java:5017)
	at com.facebook.react.uimanager.ViewGroupManager.addView(ViewGroupManager.java:38)
	at com.facebook.react.uimanager.NativeViewHierarchyManager.manageChildren(NativeViewHierarchyManager.java:533)
	at com.swmansion.reanimated.layoutReanimation.ReanimatedNativeHierarchyManager.manageChildren(ReanimatedNativeHierarchyManager.java:300)
	at com.facebook.react.uimanager.UIViewOperationQueue$ManageChildrenOperation.execute(UIViewOperationQueue.java:217)
	at com.facebook.react.uimanager.UIViewOperationQueue$1.run(UIViewOperationQueue.java:915)
	at com.facebook.react.uimanager.UIViewOperationQueue.flushPendingBatches(UIViewOperationQueue.java:1026)
	at com.facebook.react.uimanager.UIViewOperationQueue.-$$Nest$mflushPendingBatches(Unknown Source:0)
	at com.facebook.react.uimanager.UIViewOperationQueue$DispatchUIFrameCallback.doFrameGuarded(UIViewOperationQueue.java:1086)
	at com.facebook.react.uimanager.GuardedFrameCallback.doFrame(GuardedFrameCallback.java:29)
	at com.facebook.react.modules.core.ReactChoreographer$ReactChoreographerDispatcher.doFrame(ReactChoreographer.java:175)
	at com.facebook.react.modules.core.ChoreographerCompat$FrameCallback$1.doFrame(ChoreographerCompat.java:85)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1229)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1239)
	at android.view.Choreographer.doCallbacks(Choreographer.java:899)
	at android.view.Choreographer.doFrame(Choreographer.java:827)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1214)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7918)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
```

**Comment by jacobp100** [comment]
Do you have a few traces before and after that?

**Comment by mgerdes** [comment]
I'm not sure what you mean? Before and after it's inside `runOnNativeModulesQueueThread`? Or still inside of `runOnNativeModulesQueueThread`?

**Comment by jacobp100** [comment]
I think some other functions - presumably outside of this library - will be calling `runOnNativeModulesQueueThread` before or after this call. If there are, what are they?

**Comment by smokfyz** [comment]
Hi. I got the same issue on iOS

**Comment by enahum** [comment]
facing the same problem after updating to RN 0.72+ including 0.73-alpha(s).. this is in combination with `react-native-navigation` from wix.

**Comment by danilobuerger** [comment]
Experiencing the same issue after upgrading to RN 0.72 (from 0.68). @jacobp100 is there anything I can do to help you debug this?

**Comment by jacobp100** [comment]
Somebody needs to look into it. There’s only two people that actively maintain this, and I don’t think either of us will investigate this - so if you want this fixed, do take a look for yourself. We are of course on hand to answer any questions etc.

**Comment by danilobuerger** [comment]
So what I figured out is that 
```
uiManager.uiImplementation.dispatchViewUpdates(-1)
```
scheduled on the native module queue is called after the lock times out. I would assume that this means that `waitForReactLayout` blocks the execution of anything on the native modules queue.

**Comment by danilobuerger** [comment]
@jacobp100 I found the offending code 🥳 

In `SafeAreaView` we call 
```
uiManager.setViewLocalData(id, localData)
```
before
```
waitForReactLayout()
```

This runs 
```
mUIImplementation.setViewLocalData(tag, data);
```
on the native modules thread.

This in turn calls
```
dispatchViewUpdatesIfNeeded() -> dispatchViewUpdates()
```

Since rn 72 (https://github.com/facebook/react-native/commit/bc766ec7f8b18ddc0ff72a2fff5783eeeff24857) we have 
```
if (getRootViewNum() <= 0) {
  // If there are no RootViews registered, there will be no View updates to dispatch.
  // This is a hack to prevent this from being called when Fabric is used everywhere.
  // This should no longer be necessary in Bridgeless Mode.
  return;
}
```
guarding the entry and this in turn waits for the main thread where we are already waiting for this thread. Deadlock.

**Comment by jacobp100** [comment]
🎉 do you know what the fix is?

**Comment by danilobuerger** [comment]
Sadly no idea yet. The problem is that `waitForReactLayout` is called inside `NativeViewHierarchyManager.manageChildren` (see stack trace above) and that synchronizes on the same object as `getRootViewNum` so it will always deadlock. Maybe @feiyin0719 or @kelset from the original PR introducing this know more 🤷‍♂️

**Comment by jacobp100** [comment]
My experience with core dev is they’re too busy to get back to you. If you find a work around though, we here are good at merging and releasing PRs

**Comment by jacobp100** [comment]
https://github.com/th3rdwave/react-native-safe-area-context/blob/f7e7e20b930c1e414ea34ede5d7aa1c406edc36a/android/src/main/java/com/th3rdwave/safeareacontext/SafeAreaView.kt#L53-L60

> Sadly there doesn't seem to be a way to properly dirty a yoga node from java

I'm not certain this is the case - see this:

https://github.com/facebook/react-native/blob/b5e08e80d90b6d03d1f49f0674c01f03ee300c46/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextShadowNode.java#L318-L319C14

I think it has to happen in the shadow node, so might need a slight refactor. But maybe this could avoid the deadlock?

**Comment by feiyin0719** [comment]
> @jacobp100 I found the offending code 🥳
> 
> In `SafeAreaView` we call
> 
> ```
> uiManager.setViewLocalData(id, localData)
> ```
> 
> before
> 
> ```
> waitForReactLayout()
> ```
> 
> This runs
> 
> ```
> mUIImplementation.setViewLocalData(tag, data);
> ```
> 
> on the native modules thread.
> 
> This in turn calls
> 
> ```
> dispatchViewUpdatesIfNeeded() -> dispatchViewUpdates()
> ```
> 
> Since rn 72 ([facebook/react-native@bc766ec](https://github.com/facebook/react-native/commit/bc766ec7f8b18ddc0ff72a2fff5783eeeff24857)) we have
> 
> ```
> if (getRootViewNum() <= 0) {
>   // If there are no RootViews registered, there will be no View updates to dispatch.
>   // This is a hack to prevent this from being called when Fabric is used everywhere.
>   // This should no longer be necessary in Bridgeless Mode.
>   return;
> }
> ```
> 
> guarding the entry and this in turn waits for the main thread where we are already waiting for this thread. Deadlock.

maybe we can move getRootViewNum to UIManager.onBatchComplete to let it run as before

**Comment by danilobuerger** [comment]
@feiyin0719 Thanks for getting back to me! I just tested your suggestion and it does work in this case, we no longer deadlock. I am unsure if it still fixes your memory leak. Should I push a branch so you can validate?

**Comment by kelset** [comment]
hey folks - I'm currently not on the release crew and just got back from my holiday break; do you still need me to help somehow? Looks like a commit cherry picked in a 0.72 patch was the reason why this problem started happening, am I following correctly?

**Comment by Svarto** [comment]
@danilobuerger I have the same issue in my app, so a fix would be fantastic. Not sure what about the memory leak problem you refer to - can't find that discussed yet in this thread...

**Comment by jacobp100** [comment]
@kelset yes, but I don't have enough context about the change

**Comment by kelset** [comment]
if that's the case I'd suggest to flag up this problem in the latest open discussion on 0.72 patch and ask for the cherry pick to be reverted: https://github.com/reactwg/react-native-releases/discussions/97

**Comment by Bowlerr** [comment]
Hey, is there any patch for this temporarily?

**Comment by vlanemcev** [comment]
@danilobuerger Hi!

Were you able to fix the issue by making some patch-package or anything else?

**Comment by vlanemcev** [comment]
To everyone who is needed temporary fix. (@enahum @OrlovMaks )

In our project we have an version for IOS with `SafeAreaView` usage: 

```
import React, { memo, useMemo } from 'react';
import {
  Edge,
  NativeSafeAreaViewProps,
  SafeAreaView as SafeAreaContextView,
} from 'react-native-safe-area-context';
import { useHeaderHeight } from '@react-navigation/elements';

import { commonFlex } from 'src/app/theme/constants/commonStyles';

import { useIsNetworkConnected } from 'src/shared/components/NetworkConnection';
import { isProduction } from 'src/shared/constants/environment';

type SafeAreaViewProps = {
  /**
   * Should we add top inset which will be respect Header height
   * Read more about this specific behaviour:
   * https://reactnavigation.org/docs/nesting-navigators#parent-navigators-ui-is-rendered-on-top-of-child-navigator
   * https://reactnavigation.org/docs/elements#headertransparent
   */
  shouldAddHeaderHeightInset?: boolean;
} & NativeSafeAreaViewProps;

const SafeAreaView = ({
  shouldAddHeaderHeightInset,
  edges = [],
  style,
  children,
  ...rest
}: SafeAreaViewProps) => {
  const parentScreenHeaderHeight = useHeaderHeight();

  const isNetworkConnected = useIsNetworkConnected();

  const isAnyAppStatuesShown = !isProduction || !isNetworkConnected;

  // Remove `bottom` edge from SafeAreaContextView if AppMetaInfoBar or NetworkConnectionBar is displayed.
  const computedEdges = useMemo(
    () =>
      (edges as Edge[]).filter((edge) => {
        return isAnyAppStatuesShown ? edge !== 'bottom' : edge;
      }),
    [edges, isAnyAppStatuesShown]
  );

  return (
    <SafeAreaContextView
      edges={['left', 'right', ...computedEdges]}
      style={[
        commonFlex.flexContainer,
        shouldAddHeaderHeightInset
          ? { paddingTop: parentScreenHeaderHeight }
          : {},
        style,
      ]}
      {...rest}
    >
      {children}
    </SafeAreaContextView>
  );
};

export default memo(SafeAreaView);
```

and dedicated version for Android (to mitigate this issue), which `useSafeAreaInsets` usage: 

```
import React, { memo, useMemo } from 'react';
import { View, ViewProps } from 'react-native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import { useHeaderHeight } from '@react-navigation/elements';

import { commonFlex } from 'src/app/theme/constants/commonStyles';

import { useIsNetworkConnected } from 'src/shared/components/NetworkConnection';
import { isProduction } from 'src/shared/constants/environment';

// TODO: watch this issue: https://github.com/th3rdwave/react-native-safe-area-context/issues/448
// to come back to the previous realisation of SafeAreaView (as for IOS)

interface SafeAreaViewProps extends ViewProps {
  edges?: ('top' | 'right' | 'bottom' | 'left')[]; // edges as an array of strings
  children?: React.ReactNode;

  /**
   * Should we add top inset which will be respect Header height
   * Read more about this specific behaviour:
   * https://reactnavigation.org/docs/nesting-navigators#parent-navigators-ui-is-rendered-on-top-of-child-navigator
   * https://reactnavigation.org/docs/elements#headertransparent
   */
  shouldAddHeaderHeightInset?: boolean;
}

const SafeAreaView = ({
  shouldAddHeaderHeightInset,
  edges = [],
  style,
  children,
  ...rest
}: SafeAreaViewProps) => {
  const safeAreaInsets = useSafeAreaInsets();
  const parentScreenHeaderHeight = useHeaderHeight();

  const isNetworkConnected = useIsNetworkConnected();
  const isAnyAppStatuesShown = !isProduction || !isNetworkConnected;

  const computedEdges = useMemo(() => {
    // Include 'left' and 'right' by default
    let finalEdges = ['left', 'right', ...edges];

    if (isAnyAppStatuesShown) {
      // Remove `bottom` edge from SafeAreaContextView if AppMetaInfoBar or NetworkConnectionBar is displayed.

      finalEdges = finalEdges.filter((edge) => edge !== 'bottom');
    }

    return finalEdges;
  }, [edges, isAnyAppStatuesShown]);

  const computedPadding = useMemo<{
    paddingTop?: number;
    paddingBottom?: number;
    paddingLeft?: number;
    paddingRight?: number;
  }>(() => {
    let padding = {};

    if (computedEdges.includes('top')) {
      padding = { ...padding, paddingTop: safeAreaInsets.top };
    }

    if (computedEdges.includes('bottom')) {
      padding = { ...padding, paddingBottom: safeAreaInsets.bottom };
    }

    if (computedEdges.includes('left')) {
      padding = { ...padding, paddingLeft: safeAreaInsets.left };
    }

    if (computedEdges.includes('right')) {
      padding = { ...padding, paddingRight: safeAreaInsets.right };
    }

    return padding;
  }, [computedEdges, safeAreaInsets]);

  // Calculate the total top padding
  const topPadding = shouldAddHeaderHeightInset
    ? (computedPadding?.paddingTop ?? 0) + parentScreenHeaderHeight
    : computedPadding?.paddingTop;

  return (
    <View
      style={[
        commonFlex.flexContainer,
        { ...computedPadding, paddingTop: topPadding },
        style,
      ]}
      {...rest}
    >
      {children}
    </View>
  );
};

export default memo(SafeAreaView);
```

**Comment by chaadow** [comment]
I think I found a universal solution ( and it worked for my case) . I followed this [documentation](https://reactnavigation.org/docs/handling-safe-area/) and it recommends to use `useSafeAreaInsets` instead of `SafeAreaView`.
<img width="858" alt="image" src="https://github.com/th3rdwave/react-native-safe-area-context/assets/1371733/e403b449-ef58-4b2e-b777-bf8069b24158">


All I did was apply the recommandation ( by using a **regular View with paddings applied** and  no more jumpy or freezing for me!

 Here is an example

```tsx
function Demo() {
  const insets = useSafeAreaInsets();

  return (
    <View
      style={{
        // Paddings to handle safe area
        paddingTop: insets.top,
        paddingBottom: insets.bottom,
        paddingLeft: insets.left,
        paddingRight: insets.right,
      }}
    >
      <Text>Some text.</Text>
    </View>
  );
}
```

**Comment by yayvery** [comment]
The previous two "fixes" suggesting to use `useSafeAreaInsets` instead of `SafeAreaView` are missing the point of why `SafeAreaView` exists.

**Comment by chaadow** [comment]
@yayvery I'm simply following the official documentation recommendation and it did fix the issue for me ( I've linked it + provided a screenshot of the relevant section) 

If we're both missing the point, please enlighten us all as well as the official maintainers of the package how to use SafeAreaView in this context.

**Comment by yayvery** [comment]
If it works for you that's great, but it's misleading to call it a "universal solution" when it doesn't at all address the actual issue that has been identified (https://github.com/th3rdwave/react-native-safe-area-context/issues/448#issuecomment-1871155936)

The point of SafeAreaView is clearly explained in the README.

**Comment by EvertEt** [comment]
> @feiyin0719 Thanks for getting back to me! I just tested your suggestion and it does work in this case, we no longer deadlock. I am unsure if it still fixes your memory leak. Should I push a branch so you can validate?

@feiyin0719 Is there any way we can support testing this change to get it merged back? It looks like this issue affects several apps. Thanks in advance

**Comment by patrickkempff** [comment]
We have the same problem, we are trying to apply the patch from https://github.com/discord/react-native/commit/5c0d4da399050da583cd31bf2620fea9a2a30cf1 to see if it fixes our issue.

**Comment by vanHoi** [comment]
We also have this issue in our app. When applying the patch from discord/react-native@5c0d4da and building React Native from source, it fixes the issue.

**Comment by yayvery** [comment]
thanks for confirming it fixes the issue @vanHoi ! I'll try to get this reviewed & merged into react-native (need to figure out the CLA situation tho)

**Comment by EvertEt** [comment]
Thanks @yayvery , let us know if you need help or want someone else to open the PR.

**Comment by vanHoi** [comment]
Thanks @yayvery for the solution! Just to clarify, we have applied this patch to React Native 0.73.6, the issues still persists there.

**Comment by EvertEt** [comment]
I can also confirm the patch works on 0.72.12.

**Comment by jackylu0124** [comment]
@yayvery @EvertEt @vanHoi hey all, I have a quick question, is this fix incorporated into the React Native 0.72.12 version? If not, is there a plan to incorporate this fix into the next 0.72 patch version? Thanks a lot in advance!

I think this issue (https://github.com/th3rdwave/react-native-safe-area-context/issues/219#issuecomment-1913104995) is also related to one discussed above.

**Comment by EvertEt** [comment]
> @yayvery @EvertEt @vanHoi hey all, I have a quick question, is this fix incorporated into the React Native 0.72.12 version? If not, is there a plan to incorporate this fix into the next 0.72 patch version? Thanks a lot in advance!
> 
> I think this issue (https://github.com/th3rdwave/react-native-safe-area-context/issues/219#issuecomment-1913104995) is also related to one discussed above.

@jackylu0124 the fix has not been merged in react native yet, and thus also not yet backported to 0.72.

**Comment by jackylu0124** [comment]
> > @yayvery @EvertEt @vanHoi hey all, I have a quick question, is this fix incorporated into the React Native 0.72.12 version? If not, is there a plan to incorporate this fix into the next 0.72 patch version? Thanks a lot in advance!
> > I think this issue ([#219 (comment)](https://github.com/th3rdwave/react-native-safe-area-context/issues/219#issuecomment-1913104995)) is also related to one discussed above.
> 
> @jackylu0124 the fix has not been merged in react native yet, and thus also not yet backported to 0.72.

@EvertEt I see, thank you for the clarification! Is there a plan/or a task already made for merging the fix into React Native and backporting it to 0.72? Or are we not at that step yet?

**Comment by EvertEt** [comment]
> thanks for confirming it fixes the issue @vanHoi ! I'll try to get this reviewed & merged into react-native (need to figure out the CLA situation tho)

@jackylu0124, @yayvery was going to look into this but not sure what the timeline would be.

**Comment by EvertEt** [comment]
@feiyin0719 @yayvery Since this is quite a big issue for our project, I've gone ahead and opened a PR https://github.com/facebook/react-native/pull/43643 describing the issue as best as I could. Comments and possible improvements would be much appreciated.

**Comment by ziga-hvalec** [comment]
this does not fix the issue for me. 

This is my patch file. Did I miss something?

```
diff --git a/node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIImplementation.java b/node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIImplementation.java
index 894f323..0363c73 100644
--- a/node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIImplementation.java
+++ b/node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIImplementation.java
@@ -177,7 +177,7 @@ public class UIImplementation {
    *
    * @return The num of root view
    */
-  private int getRootViewNum() {
+  public int getRootViewNum() {
     return mOperationsQueue.getNativeViewHierarchyManager().getRootViewNum();
   }
 
@@ -610,12 +610,6 @@ public class UIImplementation {
 
   /** Invoked at the end of the transaction to commit any updates to the node hierarchy. */
   public void dispatchViewUpdates(int batchId) {
-    if (getRootViewNum() <= 0) {
-      // If there are no RootViews registered, there will be no View updates to dispatch.
-      // This is a hack to prevent this from being called when Fabric is used everywhere.
-      // This should no longer be necessary in Bridgeless Mode.
-      return;
-    }
     SystraceMessage.beginSection(
             Systrace.TRACE_TAG_REACT_JAVA_BRIDGE, "UIImplementation.dispatchViewUpdates")
         .arg("batchId", batchId)
diff --git a/node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java b/node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java
index 9ff04bc..2f863ad 100644
--- a/node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java
+++ b/node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java
@@ -719,7 +719,12 @@ public class UIManagerModule extends ReactContextBaseJavaModule
       listener.willDispatchViewUpdates(this);
     }
     try {
-      mUIImplementation.dispatchViewUpdates(batchId);
+      // If there are no RootViews registered, there will be no View updates to dispatch.
+      // This is a hack to prevent this from being called when Fabric is used everywhere.
+      // This should no longer be necessary in Bridgeless Mode.
+      if (mUIImplementation.getRootViewNum() > 0) {
+        mUIImplementation.dispatchViewUpdates(batchId);
+      }
     } finally {
       Systrace.endSection(Systrace.TRACE_TAG_REACT_JAVA_BRIDGE);
     }
 ```

**Comment by EvertEt** [comment]
@ziga-hvalec Make sure to also build React Native from the source: https://reactnative.dev/contributing/how-to-build-from-source

**Comment by matt-dalton** [comment]
So do we know which React Native versions (if any) include the fix?

**Comment by ziga-hvalec** [comment]
Should be 0.73.7.

**Comment by matt-dalton** [comment]
Amazing..thanks for confirming

**Comment by EvertEt** [comment]
A fix for this has been merged in v0.72.13, v0.73.7 and v0.74.0-rc8. 
The issue can be closed now here IMO.

**Comment by derrh** [comment]
We are still getting reports of ANR issues after releasing this patch (React Native 0.72.13). Or is this perhaps a different issue?

```
tb.b
	at jdk.internal.misc.Unsafe.park(Native Method)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:252)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:1672)
	at com.th3rdwave.safeareacontext.k.i(SafeAreaView.kt:47)
	at com.th3rdwave.safeareacontext.k.g(SafeAreaView.kt:75)
	at com.th3rdwave.safeareacontext.k.f(SafeAreaView.kt:24)
	at com.th3rdwave.safeareacontext.k.onAttachedToWindow(SafeAreaView.kt:21)
	at android.view.View.dispatchAttachedToWindow(View.java:20874)
	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3490)
	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3497)
	at android.view.ViewGroup.addViewInner(ViewGroup.java:5293)
	at android.view.ViewGroup.addView(ViewGroup.java:5076)
	at com.facebook.react.views.view.i.addView(ReactViewGroup.java:4)
	at android.view.ViewGroup.addView(ViewGroup.java:5016)
	at com.facebook.react.views.view.ReactClippingViewManager.addView(ReactClippingViewManager.java:6)
	at com.facebook.react.views.view.ReactClippingViewManager.addView(ReactClippingViewManager.java:2)
	at com.facebook.react.uimanager.NativeViewHierarchyManager.manageChildren(NativeViewHierarchyManager.java:417)
	at com.swmansion.reanimated.layoutReanimation.ReanimatedNativeHierarchyManager.manageChildren(ReanimatedNativeHierarchyManager.java:8)
	at com.facebook.react.uimanager.UIViewOperationQueue$ManageChildrenOperation.execute(UIViewOperationQueue.java:15)
	at com.facebook.react.uimanager.UIViewOperationQueue$1.run(UIViewOperationQueue.java:139)
	at com.facebook.react.uimanager.UIViewOperationQueue.flushPendingBatches(UIViewOperationQueue.java:54)
	at com.facebook.react.uimanager.UIViewOperationQueue.x(UIViewOperationQueue.java:1)
	at com.facebook.react.uimanager.UIViewOperationQueue$DispatchUIFrameCallback.doFrameGuarded(UIViewOperationQueue.java:32)
	at com.facebook.react.uimanager.GuardedFrameCallback.doFrame(GuardedFrameCallback.java:1)
	at com.facebook.react.modules.core.ReactChoreographer$ReactChoreographerDispatcher.doFrame(ReactChoreographer.java:47)
	at com.facebook.react.modules.core.ChoreographerCompat$FrameCallback$1.doFrame(ChoreographerCompat.java:3)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1035)
	at android.view.Choreographer.doCallbacks(Choreographer.java:845)
	at android.view.Choreographer.doFrame(Choreographer.java:775)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1022)
	at android.os.Handler.handleCallback(Handler.java:980)
	at android.os.Handler.dispatchMessage(Handler.java:104)
	at android.os.Looper.loopOnce(Looper.java:238)
	at android.os.Looper.loop(Looper.java:357)
	at android.app.ActivityThread.main(ActivityThread.java:8098)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1026)
```

**Comment by AndrewTotsky** [comment]
Temporary fixed it using `import { SafeAreaView } from 'react-native-safe-area-context';` instead of `import { SafeAreaView } from 'react-native-safe-area-context'; `

**Comment by edisoni1995** [comment]
The issue continues. I faced this after an upgrade from 0.75.5 to 0.76.7 that my screens took an extra 2 seconds to open after I navigate to them. This looks like happens only for a few times and after that when navigating everything looks fine. 
I also upgraded to latest version 0.78.1 and the latest version of `react-native-safe-area-context` and it still doesn't work.

I am using something like this in my code which get's included on every screen:

```
    <View style={[styles.flex, { backgroundColor: colors.LAYOUT }]}>
      <StatusBar
        backgroundColor={statusBarColor}
        contentColor={statusBarContentColor}
        hidden={hidden}
      />
      <SafeAreaView
        style={[styles.flexZero, { backgroundColor: statusBarColor }]}
        edges={['top']}
      />
      <SafeAreaView
        style={[styles.flex, { backgroundColor }]}
        edges={
          hasBottomTabBar ? ['left', 'right'] : ['bottom', 'left', 'right']
        }
      >
        {children}
      </SafeAreaView>
    </View>
```

If I switch to `useSafeAreaInsets` then I am not facing the issue anymore

```
const {top, bottom} = useSafeAreaInsets();
  return (
    <View style={[styles.flex, { backgroundColor: colors.LAYOUT }]}>
      <StatusBar
        backgroundColor={statusBarColor}
        contentColor={statusBarContentColor}
        hidden={hidden}
      />
      <View
        style={[styles.flexZero, { backgroundColor: statusBarColor, paddingTop: top }]}
      />
      <View
        style={[styles.flex, { backgroundColor }, { paddingBottom: hasBottomTabBar ? bottom : 0 }]}
      >
        {children}
      </View>
    </View>
```

So it looks like there is an issue only when using the `SafeAreaView` component.

**Comment by TuurDutoit** [comment]
We are also still seeing issues, though it looks like the situation at least improved with the fix merged into React Native: whereas before _every_ update to a `SafeAreaView` caused a deadlock, it happens more sporadically now.

We noticed the issue because the Google Play Console reports excessive frozen frames in our app. After some profiling with Android Studio, we narrowed it down to this deadlock in `SafeAreaView`.

## Symptoms
There are a few symptoms to look for:
- The UI hangs for 500ms
- A warning is logged in logcat: `SafeAreaView: Timed out waiting for layout.`
    - You can use `adb logcat | grep SafeAreaView` to look for it
    - This comes from [here](https://github.com/AppAndFlow/react-native-safe-area-context/blob/7b2ccee94f86d6dccf904f2dc83fd8d9289fdc16/android/src/main/java/com/th3rdwave/safeareacontext/SafeAreaView.kt#L99)
- In a trace file, you'll see tasks that take 500ms (or multiples of that!) on the main and native modules threads
    - The main thread is blocked on `DispatchUI`
    - The native modules thread is blocked on `onBatchCompleteUI`
    - Sometimes, I also see "monitor contention with owner main" on the native modules thread
- Sometimes, multiple deadlocks happen during the same render cycle, causing a 1s or even 1.5s lock up. These are reported as "frozen frames" in Google Play and other performance monitoring tools

## Reproduction
We've been able to reproduce the issue quite consistently in our app:
- When starting the app, without doing anything else, we often see the logcat log appear a few times. We assume this actually delays our app startup time.
- When opening or closing the keyboard. We can see the UI locking up, as well as the logcat log

These are both on fairly busy screens, with lots of content, `ScrollView`s, several `SaveAreaView`s, and `KeyboardAvoidingView`s. For the sake of our investigation, I slimmed down one screen to just a `TextInput` and a `SafeAreaView`. The `edges` of the `SafeAreaView` changes when the keyboard shows and hides. With this setup, I still see the issue pop up, although less often than with the full screen.

I also created a minimal reproduction based on a React Native template [here](https://github.com/TuurDutoit/SafeAreaContextDeadlock). This one updates the `edges` of a `SafeAreaView` when tapping a button. There's also a continuous animation running, which seems to increase the likelihood of the issue popping up. Without it, I can still reproduce by quickly tapping the button multiple times.
I used the RN 0.75 template for this repo (as that's what we currently run on), but I've tried with 0.78 too and can still reproduce it the same way. However, it only happens when using the old architecture - with the new arch enabled, the problem disappears.

## Root cause
The root of the issue seems to be the lock (that was mentioned earlier in this thread) in [`waitForReactLayout`](https://github.com/AppAndFlow/react-native-safe-area-context/blob/7b2ccee94f86d6dccf904f2dc83fd8d9289fdc16/android/src/main/java/com/th3rdwave/safeareacontext/SafeAreaView.kt#L67). There's a comment for it that says:

> This should be safe as long as the native modules thread does not block waiting on the main thread.

Given our findings above, it appears that it's not safe to assume this can never happen. We're not 100% sure how, but in some cases a deadlock does still happen, and it appears to happen more the "busier" a screen gets. Could there be things in the RN rendering pipeline that cause the native modules thread to wait for the main thread? If such a task is scheduled right before a `SafeAreaView` update, that could cause a deadlock.

**Comment by francescotescari** [comment]
We are also experiencing the same deadlock as per message above. 
We are on:
- React native 0.76.7
- Safe area context 5.4.1
- New arch off

Some additional details on the deadlock in our case:

Main thread simplified stack
- [UIViewOperationQueue.flushPendingBatches](https://github.com/facebook/react-native/blob/1e967ca98303ff573ccfcc84dad882f0568586ac/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIViewOperationQueue.java#L979)
- [NativeViewHierarchyManager.manageChildren](https://github.com/facebook/react-native/blob/1e967ca98303ff573ccfcc84dad882f0568586ac/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyManager.java#L397) **acquires a lock on NativeViewHierarchyManager!**
- (adds the view to the native View hierarchy)
- [SafeAreaView.onAttachedToWindow](https://github.com/AppAndFlow/react-native-safe-area-context/blob/94da3e4d16c30c04fc8187ab068904353d9312b6/android/src/main/java/com/th3rdwave/safeareacontext/SafeAreaView.kt#L135)
- [SafeAreaView.waitForReactLayout](https://github.com/AppAndFlow/react-native-safe-area-context/blob/94da3e4d16c30c04fc8187ab068904353d9312b6/android/src/main/java/com/th3rdwave/safeareacontext/SafeAreaView.kt#L67) **waits for native modules thread!**

Native modules thread
- [NativeModuleRegistry.onBatchComplete](https://github.com/facebook/react-native/blob/1e967ca98303ff573ccfcc84dad882f0568586ac/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/bridge/NativeModuleRegistry.java#L114)
- [UIManagerModule.onBatchComplete](https://github.com/facebook/react-native/blob/1e967ca98303ff573ccfcc84dad882f0568586ac/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIManagerModule.java#L681)
- [NativeViewHierarchyManager.getRootViewNum](https://github.com/facebook/react-native/blob/1e967ca98303ff573ccfcc84dad882f0568586ac/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyManager.java#L692) **acquires a lock on NativeViewHierarchyManager!**

It deadlocks twice, causing a 1s deadlock time

<img width="635" alt="Image" src="https://github.com/user-attachments/assets/133a2d5d-58fb-4d91-8dc3-fe4b97991f46" />
<img width="680" alt="Image" src="https://github.com/user-attachments/assets/0a95318a-577e-4dd3-b28b-ce45e4deaaf4" />
